src/silly_sprites/event_system.hpp

```diff
@@ -0,0 +1,67 @@
+#include "events.hpp"
+#include "types.hpp"
+#include <vector>
+
+namespace sly::event {
+
+    struct EventID final {
+        friend class EventSystem;
+
+    private:
+        usize m_id;
+        EventID(usize id) : m_id{ id } {};
+
+    public:
+        EventID() : m_id{ 0 } {};
+        [[nodiscard]] auto operator<=>(EventID const&) const = default;
+    };
+
+    class EventSystem final {
+    private:
+        std::vector<std::pair<EventID, EventCallbacks>> m_handlers;
+        usize eventID = 0;
+
+        template<Event T>
+        [[nodiscard]] bool
+        is_same_handler(std::function<void(T const&)> handler, std::function<void(T const&)> handler2) {
+            return handler == handler2;
+        }
+
+    public:
+        EventSystem() = default;
+        EventSystem(EventSystem const&) = delete;
+        EventSystem(EventSystem&&) = delete;
+        EventSystem& operator=(EventSystem const&) = delete;
+        EventSystem& operator=(EventSystem&&) = delete;
+
+        template<Event T>
+        EventID add_handler(std::function<void(T const&)> handler) {
+            auto const id = EventID{ eventID++ };
+            m_handlers.push_back({ id, handler });
+            return id;
+        }
```

Comment:
I think I don't quite understand the meaning of `eventID`. When I call `add_handler()`, the counter will be increased. Doesn't that mean that each ID is unique? So it should be impossible to add two handlers with the same ID, isn't it? But then I don't understand the data structure (`std::vector<std::pair<EventID, EventCallbacks>>`).

I think I need a more detailed explanation of what `eventID` stands for. To me it seems that it is a kind of "unique handler id", but maybe I'm overlooking something?

Reply:
it is some handler id. maybe i should rename it.
you only need it to delete the handler right now.

Reply:
Ah okay. Then it should be renamed, yes. I'm still wondering if there's a better way to handle this, but I cannot think of an alternative right now. Maybe one of us will find a better way sometime....

But concerning the data structure: Since the first entry of the `std::pair` is unique now, you could change it to use an `unordered_map`. Although that wouldn't improve searching for events of the right type, but at least it would improve adding an deleting handlers.

Reply:
i could chance it yes.
event tho it does not help while depatching.
it just helps while deleting.

Reply:
Right. If dispatching becomes a real performance issue, we would need a different approach. But I think that's not important at the moment.